chainstate/test-suite/src/tests/initialization.rs

```diff
@@ -0,0 +1,134 @@
+                num_blocks.into()
+            );
+
+            // Extract the final storage and drop the test framework
+            tf.storage.clone()
+        };
+
+        // Initialize a different test framework with given storage.
+        // The test framework should pass the genesis check (panics if not)
+        let _tf = TestFramework::builder().with_storage(storage).build();
+    })
+}
+
+fn genesis_check_err(num_blocks: u64) {
+    utils::concurrency::model(move || {
+        // Two different configs with separate genesis IDs.
+        let conf0 = ChainConfigBuilder::new(ChainType::Mainnet)
+            .net_upgrades(NetUpgrades::unit_tests())
+            .genesis_unittest(common::chain::Destination::ScriptHash(Id::new(
+                [0x00; 32].into(),
+            )))
+            .build();
+        let conf1 = ChainConfigBuilder::new(ChainType::Mainnet)
+            .net_upgrades(NetUpgrades::unit_tests())
+            .genesis_unittest(common::chain::Destination::ScriptHash(Id::new(
+                [0x01; 32].into(),
+            )))
+            .build();
+        assert_ne!(conf0.genesis_block_id(), conf1.genesis_block_id());
+
+        let storage = {
+            let mut tf = TestFramework::builder().with_chain_config(conf0).build();
+
+            // Add a bunch of blocks
+            for _ in 0..num_blocks {
+                let _index = tf
+                    .make_block_builder()
+                    .build_and_process()
+                    .expect("block processing to succeed")
+                    .expect("block index to be returned");
+            }
+            // Check the number of blocks in the storage matches
+            assert_eq!(
+                tf.chainstate.get_best_block_height().unwrap(),
+                num_blocks.into()
+            );
+
+            tf.storage
+        };
+
+        // Start another chain with different genesis using the previous storage
+        let result = TestFramework::builder()
+            .with_chain_config(conf1)
+            .with_storage(storage)
+            .try_build();
+
+        // Verify the result
+        match result {
+            Err(ChainstateError::FailedToInitializeChainstate(_)) => (),
+            Err(e) => panic!("unexpected error {e:?}, chain height {num_blocks}"),
+            Ok(_) => panic!("expected chainstate initialization to fail"),
+        }
+    })
+}
+
+#[test]
+fn genesis_check_ok_empty_chain() {
+    genesis_check_ok(0);
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn genesis_check_ok_nonempty_chain(#[case] seed: Seed) {
+    genesis_check_ok(make_seedable_rng(seed).gen_range(1..100));
+}
+
+#[test]
+fn genesis_check_err_empty_chain() {
+    genesis_check_err(0);
+}
+
+#[test]
+fn genesis_check_err_height_1() {
+    genesis_check_err(1);
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn genesis_check_err_nonempty_chain(#[case] seed: Seed) {
+    genesis_check_err(make_seedable_rng(seed).gen_range(1..100));
```

Comment:
Should the range be `2..100` because the one block case is covered by a separate test? 🙃

Reply:
Doesn't hurt to check twice :-)